Ubernetes: When scheduling, spread between zones if labeled

justinsb opened the pull request:
We already spread across nodes; we modify this spreading preference to
spread across zones when nodes are labeled with zone information.

justinsb committed:
When scheduling, spread between zones if labeled

We already spread across nodes; we modify this spreading preference to
spread across zones when nodes are labeled with zone information.

justinsb commented:
cc @quinton-hoole

quinton-hoole commented on plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go:
```diff
@@ -37,11 +38,28 @@ func NewSelectorSpreadPriority(serviceLister algorithm.ServiceLister, controller
+		return ""
+	}
+
+	return region + ":" + failureDomain
```
Does this work if region or failureDomain contain this character? If not, at least worth calling out as a precondition, or perhaps even failing loudly over.

justinsb replied:
In theory there could be a poor scheduling decision if there was a conflict, though I would be very surprised if this happened in the real world.  I'm changing the delimiter to ":\x00:"; I think not including a null is a very safe assumption (particularly given the consequence is just a suboptimal zone spreading).  Nicely, the null character is not printed, so the key prints as e.g. "us-east-1::us-east-1a"

quinton-hoole commented on plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go:
```diff
@@ -37,11 +38,28 @@ func NewSelectorSpreadPriority(serviceLister algorithm.ServiceLister, controller
+	}
+
+	return region + ":" + failureDomain
+}
```
Sins of the past, but the comment below is verging on being incomprehensible.  Worth improving upon now, while it's fresh in your mind?

justinsb replied:
Rewrote it!

quinton-hoole commented on plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go:
```diff
@@ -76,9 +94,17 @@ func (s *SelectorSpread) CalculateSpreadPriority(pod *api.Pod, podLister algorit
+	countsByNodeName := map[string]int{}
 	if len(nsPods) > 0 {
 		for _, pod := range nsPods {
+			// When we are replacing a failed pod, we often see the previous deleted version
```
Nit: Surely the if statement above is redundant? Not your code, but still...

justinsb replied:
I agree; I removed it

quinton-hoole commented on plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go:
```diff
@@ -87,24 +113,62 @@ func (s *SelectorSpread) CalculateSpreadPriority(pod *api.Pod, podLister algorit
+	maxCountByZone := 0
+	haveZones := false
+	countsByZone := map[string]int{}
+	for i := range nodes.Items {
```
This iterates over all nodes a second time for each pod placement?  Surely we should rather do this as part of the first iteration above rather?

justinsb replied:
The first iteration was over the pods; now we are iterating over the nodes to find the counts-by-zone.  We could build a map of node-name to node and then combine these two loops, but we'd need a loop to build that map (and I think this is more efficient than building that node-map).

I am making this a little more efficient and perhaps easier to understand though, by computing the aggregates in a separate loop rather than in the body of the main loop.  It does mean more loops, but each loop is much simpler.

quinton-hoole commented on plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go:
```diff
@@ -87,24 +113,62 @@ func (s *SelectorSpread) CalculateSpreadPriority(pod *api.Pod, podLister algorit
 	//score int - scale of 0-10
 	// 0 being the lowest priority and 10 being the highest
-	for _, node := range nodes.Items {
+	for i := range nodes.Items {
```
This is a third iteration over all nodes?  It might be more difficult to eliminate this one, but worth trying.

justinsb replied:
Well, it's the second.  I think this loop cannot be combined, because it needs the 'max-pods-by-node' and 'max-pods-by-zone', which can only be known after every pod & every zone have been examined.

quinton-hoole commented on plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go:
```diff
@@ -87,24 +113,62 @@ func (s *SelectorSpread) CalculateSpreadPriority(pod *api.Pod, podLister algorit
-		if maxCount > 0 {
-			fScore = 10 * (float32(maxCount-counts[node.Name]) / float32(maxCount))
+		if maxCountByNodeName > 0 {
+			fScore = 10 * (float32(maxCountByNodeName-countsByNodeName[node.Name]) / float32(maxCountByNodeName))
```
Please lift this magic number 10 up into a named constant.

justinsb replied:
Done - I called it maxPriority

quinton-hoole commented on plugin/pkg/scheduler/algorithm/priorities/selector_spreading.go:
```diff
@@ -87,24 +113,62 @@ func (s *SelectorSpread) CalculateSpreadPriority(pod *api.Pod, podLister algorit
+		if haveZones {
+			zoneId := getZoneKey(node)
+			if zoneId != "" {
+				fScore += 20 * (float32(maxCountByZone-countsByZone[zoneId]) / float32(maxCountByZone))
```
Comment as above for magic number 20.

quinton-hoole replied:
Also, this is inconsistent with the 0-10 scale mentioned in the comment above, as this can allocate a max score of 20 (until divided by 3 below).  Need to make this all easier to understand, IMO.  2/3 is the zone vs node weighting, and 10 is the scoring range.  Lets make that explicit with named constants, rather than implicit via 20/3 split over multiple lines.

quinton-hoole replied:
I think that this makes the incorrect trade-off in some cases? e.g. given zone A with 3 nodes, and zone B with 1 node, it will place the 6th pod on the node in zone B (together with 2 other pods), so that losing that single node loses 50% of pods.  Losing a node is highly more probable than losing a zone, so surely the weighting should be towards node spreading rather than zone spreading? Actually, thinking about it some more, I think that a reasonable argument could be made both ways.  The bottom line is, once we have reached even numbers of pods in all zones, and one of the zones is out of capacity of "empty" nodes, is it preferable to have too many pods in one zone, or too many pods on one node?  Perhaps we need to run some simulations to decide which is a better default, but in the mean time it seems like the choice between the two strategies should at least be configurable?

justinsb replied:
I have made it configurable, but left the TODO about "any way to justify 2/3 to zone-spreading".  I agree that it isn't entirely clear which is better.  I like your argument that losing a node is more likely than losing a zone.  Another strategy might be to first ensure that we have one pod in every zone; then we ensure that we have one pod on every node; then 2 in every zone; then 2 on every node etc.  But at the same time I don't think we want to make these hard constraints; if there's no capacity in a zone, then we should still schedule an RC even if they all end up in one AZ.  (The user can do hard-zone-spreading using node-selectors if needed).

I suspect the really tricky thing comes in figuring out how this interacts with our other scheduling preference functions...

My personal preference is to try it and see how we feel about the decisions it makes.  I don't think they are bad (particularly once we ignore deleted pods!), though they are not perfect.  To me the bigger problems are around the lack of rescheduling when nodes are created, and around the 'victim pod' not being priority aware as an RC is scaled down.

quinton-hoole commented on plugin/pkg/scheduler/algorithm/priorities/selector_spreading_test.go:
```diff
@@ -228,6 +229,203 @@ func TestSelectorSpreadPriority(t *testing.T) {
+		wellknownlabels.LabelZoneFailureDomain: "zone3",
+	}
+	labeledNodes := map[string]map[string]string{
+		"machine1.zone1": nodeLabelsZone1,
```
I'm a bit worried about how many times these string literals are cut'n pasted throughout this source file.  A single constant array, or a generator function would significantly reduce the probability of these drifting out of sync.

quinton-hoole replied:
Also, I think we need a test involving unequal numbers of nodes per zone.

justinsb replied:
We definitely have unequal number of nodes per zone: zone1 has 1 node, zone2 has 2 nodes, zone3 has 3 nodes.

I replaced the nodeLabels with a generator function. I also replaced the string constants with constant variables, so the compiler will pick up any typos.  I'm happy to do anything better but I can't see what that would be - the problem is that some of the later test cases have some pretty complicated behavior.

quinton-hoole commented:
@justinsb let me know what you think

quinton-hoole commented:
Note: docs/devel/scheduler.md and docs/devel/scheduler_algorithm.md also need to be updated accordingly.

justinsb committed:
Zone-scheduler: Fixes per code-review

justinsb committed:
Zone scheduler: Update scheduler docs

There's not a huge amount of detail in the docs as to how the scheduler
actually works, which is probably a good thing both for readability and
because it makes it easier to tweak the zone-spreading approach in the
future, but we should include some information that we do spread across
zones if zone information is present on the nodes.

justinsb commented:
Made changes per code review, and added some hints to the docs that we now try to spread across zones also.

quinton-hoole commented:
LGTM!

quinton-hoole commented:
@k8s-bot unit test this please

quinton-hoole commented:
@k8s-bot unit test this

quinton-hoole commented:
Looks like a unit test flake. 
<testsuite tests="9" failures="1" time="1.451" name="k8s.io/kubernetes/pkg/storage">
<properties>
<property name="go.version" value="go1.4.3"/>
</properties>
<testcase classname="storage" name="TestEtcdParseWatchResourceVersion" time="0.000"/>
<testcase classname="storage" name="TestWatchCacheBasic" time="0.000"/>
<testcase classname="storage" name="TestEvents" time="0.000"/>
<testcase classname="storage" name="TestWaitUntilFreshAndList" time="0.000"/>
<testcase classname="storage" name="TestReflectorForWatchCache" time="0.000"/>
<testcase classname="storage" name="TestList" time="0.270"/>
<testcase classname="storage" name="TestWatch" time="0.370"/>
<testcase classname="storage" name="TestWatcherTimeout" time="0.160">
<failure message="Failed" type="">
cacher_test.go:235: Unexpected error: too old resource version: 2 (3)
</failure>
</testcase>
<testcase classname="storage" name="TestFiltering" time="0.380"/>
</testsuite>

@k8s-bot unit test this
